staging/src/k8s.io/component-base/metrics/prometheusextension/timing_histogram.go

```diff
@@ -0,0 +1,190 @@
+	Help        string
+	ConstLabels prometheus.Labels
+
+	// Buckets defines the buckets into which observations are
+	// accumulated. Each element in the slice is the upper
+	// inclusive bound of a bucket. The values must be sorted in
+	// strictly increasing order. There is no need to add a
+	// highest bucket with +Inf bound. The default value is
+	// prometheus.DefBuckets.
+	Buckets []float64
+
+	// The initial value of the variable.
+	InitialValue float64
+}
+
+// NewTimingHistogram creates a new TimingHistogram
+func NewTimingHistogram(opts TimingHistogramOpts) (TimingHistogram, error) {
+	return NewTestableTimingHistogram(clock.RealClock{}, opts)
+}
+
+// NewTestableTimingHistogram creates a TimingHistogram that uses a mockable clock
+func NewTestableTimingHistogram(clock clock.PassiveClock, opts TimingHistogramOpts) (TimingHistogram, error) {
+	desc := prometheus.NewDesc(
+		prometheus.BuildFQName(opts.Namespace, opts.Subsystem, opts.Name),
+		opts.Help,
+		nil,
+		opts.ConstLabels,
+	)
+	return newTimingHistogram(clock, desc, opts)
+}
+
+func newTimingHistogram(clock clock.PassiveClock, desc *prometheus.Desc, opts TimingHistogramOpts, variableLabelValues ...string) (TimingHistogram, error) {
+	allLabelsM := prometheus.Labels{}
+	allLabelsS := prometheus.MakeLabelPairs(desc, variableLabelValues)
+	for _, pair := range allLabelsS {
+		if pair == nil || pair.Name == nil || pair.Value == nil {
+			return nil, errors.New("prometheus.MakeLabelPairs returned a nil")
+		}
+		allLabelsM[*pair.Name] = *pair.Value
+	}
+	weighted, err := NewWeightedHistogram(WeightedHistogramOpts{
+		Namespace:   opts.Namespace,
+		Subsystem:   opts.Subsystem,
+		Name:        opts.Name,
+		Help:        opts.Help,
+		ConstLabels: allLabelsM,
+		Buckets:     opts.Buckets,
+	})
+	if err != nil {
+		return nil, err
+	}
+	return &timingHistogram{
+		clock:       clock,
+		weighted:    weighted,
+		lastSetTime: clock.Now(),
+		value:       opts.InitialValue,
+	}, nil
+}
+
+type timingHistogram struct {
+	clock    clock.PassiveClock
+	weighted WeightedHistogram
+
+	lock sync.Mutex // applies to all the following
+
+	// identifies when value was last set
+	lastSetTime time.Time
+	value       float64
+}
+
+var _ TimingHistogram = &timingHistogram{}
+
+func (th *timingHistogram) Set(newValue float64) {
+	th.update(func(float64) float64 { return newValue })
+}
+
+func (th *timingHistogram) Inc() {
+	th.update(func(oldValue float64) float64 { return oldValue + 1 })
+}
+
+func (th *timingHistogram) Dec() {
+	th.update(func(oldValue float64) float64 { return oldValue - 1 })
+}
+
+func (th *timingHistogram) Add(delta float64) {
+	th.update(func(oldValue float64) float64 { return oldValue + delta })
+}
+
+func (th *timingHistogram) Sub(delta float64) {
+	th.update(func(oldValue float64) float64 { return oldValue - delta })
+}
+
+func (th *timingHistogram) SetToCurrentTime() {
+	th.update(func(oldValue float64) float64 { return th.clock.Since(time.Unix(0, 0)).Seconds() })
```

Comment:
I'm really confused by this method, are you intending to set the gauge value to the unix timestamp or are you intending to set the `th.lastSetTime` to the current timestamp?

Reply:
I think this method will just set the current time as the new value it will not override the `lastSetTime`. At least that would be the behavior of the normal `SetToCurrentTime` method of a gauge. I find it also a bit confusing, but I guess it was a shortcut made in client_golang to simplify the use of gauge with timestamp metrics.

Reply:
This is strange to me too, but effectively this method is coming from the original prometheus interface:
https://pkg.go.dev/github.com/prometheus/client_golang/prometheus#Gauge

```
// SetToCurrentTime sets the Gauge to the current Unix time in seconds.
SetToCurrentTime()
```

So it works as designed :)

Reply:
I created `GaugeOps` from the existing `Gauge` interface, just removing the inclusion of Metric and Collector.  Frankly, prometheus should have this interface (analogous to the way it has `Observer = Histogram - {Metric, Collector}`).